### PR TITLE
Issue #5850 support Accelerate your network share hosted Package Source

### PR DESCRIPTION
**Pull request by nschwerzler**
Link: https://github.com/NuGet/Home/issues/5850
 Regression: No

Fix: Added functionality to nuget.exe pack command to support network share hosted package source requirements

Details:

Added two arguments to the pack command to support Accelerate your network share hosted Package Source
- OutputSHA512PackageHash: Creates the [packageName].nupkg.sha512 file that contains the package hash
- OutputResolvedNuSpec: Writes the resolved nuspec file to the package output directory which contains the same contents as the nuspec in the nupkg. If you attempt to use this option where the input nuspec is the same as the output resolved nuspec, will throw so user does not overwrite the original accidently.

Testing/Validation

Tests Added: Yes

Added 3 NuGetPackCommandTests
- PackCommand_OutputResolvedNuSpecFileAttemptToOverwriteOriginal: Verifies nuget.exe pack will not overwrite the users original nuspec when writing a resolved nuspec file to the package output directory
- PackCommand_OutputResolvedNuSpecFile: Verifies nuget.exe pack will output the resolved nuspec file with token replacements and is identical as the one in the nupkg
- PackCommand_OutputSHA512PackageHash: Verifies nuget.exe pack will output a sha512 package hash file

 Validation done:
Verified new added tests pass inside VS test explorer UI
Built product in visual studio and build.ps1 and verified passed
Looked at test results output for build.ps1 and verified tests passed

## Bug
Link: Link_to_issue  
Regression: Yes/No  
If Regression then when did it last work:   
If Regression then how are we preventing it in future:   

## Fix
Details: Details_about the_fix  

## Testing/Validation
Tests Added: Yes/No  
Reason for not adding tests:  
Validation done:

### CODE REVIEW

File: src/NuGet.Clients/NuGet.CommandLine/Commands/PackCommand.cs [hunk 1]
```diff
@@ -79,6 +79,12 @@ public ICollection<string> Exclude
         [Option(typeof(NuGetCommand), "CommandMSBuildPath")]
         public string MSBuildPath { get; set; }
 
+        [Option(typeof(NuGetCommand), "PackageCommandOutputResolvedNuSpec")]
+        public bool OutputResolvedNuSpec { get; set; }
+
+        [Option(typeof(NuGetCommand), "PackageCommandOutputSHA512PackageHash")]
+        public bool OutputSHA512PackageHash { get; set; }
```
Comment (emgarten) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/PackCommand.cs]: Should these flags be combined? Or should this be a config option instead? 

I'm worried about having too many knobs on pack, there are already a lot of options.

Comment (nschwerzler) [reply src/NuGet.Clients/NuGet.CommandLine/Commands/PackCommand.cs]: We could combine them. As far as I know NuGet pack does not support the config file option.

Comment (emgarten) [reply src/NuGet.Clients/NuGet.CommandLine/Commands/PackCommand.cs]: Pack could take the config file option, I'm not sure that is the best place for it, just an option. 
@rohit21agrawal

Comment (rohit21agrawal) [reply src/NuGet.Clients/NuGet.CommandLine/Commands/PackCommand.cs]: i'd rather have an option like -InstallPackageToV3Feed which takes the path to the feed as a value.

File: src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/PackageBuilder.cs [hunk 2]
```diff
@@ -314,6 +314,15 @@ public Version MinClientVersion
             set;
         }
 
+        /// <summary>
+        /// Gets the package manifest which is available after WriteManifest
+        /// </summary>
+        public Manifest PackageManifest
```
Comment (rohit21agrawal) [inline src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/PackageBuilder.cs]: why do you need this as a property?

File: src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/Manifest.cs [hunk 3]
```diff
@@ -51,6 +51,11 @@ public Manifest(ManifestMetadata metadata, ICollection<ManifestFile> files)
         public bool HasFilesNode { get; }
 
         /// <summary>
+        /// Gets the resolved manifest XDocument populated after save
+        /// </summary>
+        public XDocument ResolvedManifest { get; private set; }
```
Comment (rohit21agrawal) [inline src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/Manifest.cs]: same here, why is this needed as a property?

**Comment by rohit21agrawal** [comment]
if we are doing this, we also need to do feature parity for dotnet.exe pack as well. @rrelyea  thoughts?

File: src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs [hunk 4]
```diff
@@ -131,10 +131,69 @@ public PackageArchiveReader BuildPackage(PackageBuilder builder, string outputPa
...
         }
 
+        /// <summary>
+        /// Writes the resolved NuSpec file to the package output directory.
+        /// </summary>
+        /// <param name="packageOutputPath">The output package path</param>
+        /// <param name="packageBuilder">The package builder</param>
+        private void WriteResolvedNuSpecToPackageOutputDirectory(string packageOutputPath, PackageBuilder packageBuilder)
+        {
+            string packageOutputDirectory = Path.GetDirectoryName(packageOutputPath);
```
Comment (rohit21agrawal) [inline src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs:152]: we are doing something similar here:
https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Commands/MSBuildProjectFactory.cs#L90-L108

i'd rather have a similar implementation at both places. you just need to pass null for the FileCollection to Manifest constructor

File: src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs [hunk 5]
```diff
@@ -131,10 +131,69 @@ public PackageArchiveReader BuildPackage(PackageBuilder builder, string outputPa
...
+        }
+
+        /// <summary>
+        /// Writes the sha512 package hash file to the package output directory
+        /// </summary>
+        /// <param name="packageOutputPath">The output package path</param>
+        /// <param name="packageBuilder">The package builder</param>
+        private void WriteSHA512PackageHash(string packageOutputPath, PackageBuilder packageBuilder)
+        {
+            string packageOutputDirectory = Path.GetDirectoryName(packageOutputPath);
```
Comment (rohit21agrawal) [inline src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs]: @emgarten there must already be a helper function that does this , right? in install/restore? can you point nick to it?

**Comment by rohit21agrawal** [comment]
i would also like to discuss how to handle symbols nupkgs here.

**Review by rohit21agrawal — CHANGES_REQUESTED** [review]
left comments for feedback

File: src/NuGet.Clients/NuGet.CommandLine/Commands/PackCommand.cs (hunk 1 above)
Comment (nschwerzler) [reply src/NuGet.Clients/NuGet.CommandLine/Commands/PackCommand.cs]: Will consolidate to one option

File: src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs (hunk 4 above)
Comment (nschwerzler) [reply src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs:152]: Nice, thank you for link. Updating to use the same implementation in the linked code.

File: src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/Manifest.cs (hunk 3 above)
Comment (nschwerzler) [reply src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/Manifest.cs]: Removing, not needed anymore

File: src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/PackageBuilder.cs (hunk 2 above)
Comment (nschwerzler) [reply src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/PackageBuilder.cs]: Removing, not needed anymore

File: src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs (hunk 4 above)
Comment (nschwerzler) [reply src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs:152]: Resolved

File: src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs [hunk 6]
```diff
@@ -131,10 +131,63 @@ public PackageArchiveReader BuildPackage(PackageBuilder builder, string outputPa
...
 
+        /// <summary>
+        /// Writes the resolved NuSpec file to the package output directory.
+        /// </summary>
+        /// <param name="packageOutputPath">The output package path</param>
+        /// <param name="packageBuilder">The package builder</param>
+        private void WriteResolvedNuSpecToPackageOutputDirectory(string packageOutputPath, PackageBuilder packageBuilder)
+        {
+            string packageOutputDirectory = Path.GetDirectoryName(packageOutputPath);
+            string resolvedNuSpecOutputPath = Path.Combine(packageOutputDirectory, packageBuilder.Id + PackagingConstants.ManifestExtension);
```
Comment (rohit21agrawal) [inline src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs:153]: could you use this to get outputPath: https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs#L978 
modify the function to accept a parameter which decides whether or not to append version. you can use it for the sha512 file too

File: src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs [hunk 7]
```diff
@@ -131,10 +131,63 @@ public PackageArchiveReader BuildPackage(PackageBuilder builder, string outputPa
...
+            {
+                throw new PackagingException(
+                    string.Format(
+                        CultureInfo.CurrentCulture,
+                        "Unable to output resolved nuspec file because it would overwrite the original at '{0}'",
+                        _packArgs.Path));
+            }
+
+            Manifest manifest = new Manifest(new ManifestMetadata(packageBuilder), null);
+            using (Stream stream = new FileStream(resolvedNuSpecOutputPath, FileMode.Create))
```
Comment (rohit21agrawal) [inline src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs]: have you checked behavior about what happens if file already exists?

File: src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs [hunk 8]
```diff
@@ -131,10 +131,63 @@ public PackageArchiveReader BuildPackage(PackageBuilder builder, string outputPa
                 PrintVerbose(outputPath, builder);
             }
 
+            if (_packArgs.InstallPackageToV3Feed)
```
Comment (rohit21agrawal) [inline src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs]: the normal nupkg should not be generated.

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPackCommandTest.cs [hunk 9]
```diff
@@ -2085,6 +2085,181 @@ public void PackCommand_NuspecFileWithTokens()
...
+                // Verify the nuget pack command exit code
+                Assert.NotNull(commandRunner);
+                Assert.True(
+                    0 == commandRunner.Item1,
+                    string.Format("{0} {1}", commandRunner.Item2 ?? "null", commandRunner.Item3 ?? "null"));
+
+                // Verify the zip file has the resolved nuspec
+                XDocument nuspecZipXml = null;
+                var nupkgPath = Path.Combine(workingDirectory, "packageA.1.2.3.4.nupkg");
+                using (var zip = new ZipArchive(File.OpenRead(nupkgPath)))
```
Comment (rohit21agrawal) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPackCommandTest.cs]: we have PackageArchiveReader and NuspecReader to do these tasks.
https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs#L877-L880

File: src/NuGet.Core/NuGet.Commands/CommandArgs/PackArgs.cs [hunk 10]
```diff
@@ -22,6 +22,7 @@ public class PackArgs
         public ILogger Logger { get; set; }
         public LogLevel LogLevel { get; set; }
         public bool IncludeReferencedProjects { get; set; }
+        public bool InstallPackageToV3Feed { get; set; }
```
Comment (rohit21agrawal) [inline src/NuGet.Core/NuGet.Commands/CommandArgs/PackArgs.cs]: @rrelyea how about InstallPackageToOutputPath ? I propose that the flag (which is a boolean, doesn't accept any value)does the following:
1) Installs the package to the output path as specified by user
2) Prevents the normal nupkg (and symbols nupkg) from being generated.
3) Logs where the package was installed

File: src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs [hunk 11]
```diff
@@ -131,10 +131,63 @@ public PackageArchiveReader BuildPackage(PackageBuilder builder, string outputPa
...
+            {
+                WriteResolvedNuSpecToPackageOutputDirectory(outputPath, builder);
+                WriteSHA512PackageHash(outputPath, builder);
+            }
+
             _packArgs.Logger.LogMinimal(String.Format(CultureInfo.CurrentCulture, Strings.Log_PackageCommandSuccess, outputPath));
```
Comment (rohit21agrawal) [inline src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs]: the logging needs to be updated too if you are installing the package

File: src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs [hunk 12]
```diff
@@ -131,10 +131,63 @@ public PackageArchiveReader BuildPackage(PackageBuilder builder, string outputPa
...
+        {
+            string packageOutputDirectory = Path.GetDirectoryName(packageOutputPath);
+            string resolvedNuSpecOutputPath = Path.Combine(packageOutputDirectory, packageBuilder.Id + PackagingConstants.ManifestExtension);
+
+            if (string.Equals(_packArgs.Path, resolvedNuSpecOutputPath, StringComparison.OrdinalIgnoreCase))
+            {
+                throw new PackagingException(
+                    string.Format(
+                        CultureInfo.CurrentCulture,
+                        "Unable to output resolved nuspec file because it would overwrite the original at '{0}'",
```
Comment (rohit21agrawal) [inline src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs]: add this entry to the corresponding Strings.resx file so it can be localised.

**Comment by rohit21agrawal** [comment]
>             if (_packArgs.Symbols)

you also need to check if InstallPackageToOutputPath is specified and prevent this from happening

---
Refers to: src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs:739 in b2dde82. [](commit_id = b2dde829e3dd9a1ba33efadb4b6e141801638b35, deletion_comment = False)

**Comment by rohit21agrawal** [comment]
also change the implementation of PackCommandRunner.BuildSymbolsPackage so it no-ops if InstallPackageToOutputPath is set

File: src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs (hunk 11 above)
Comment (nschwerzler) [reply src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs]: Not installing the package, so this is not needed unless I am misunderstanding the request.

File: src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs (hunk 6 above)
Comment (nschwerzler) [reply src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs:153]: Resolved.

File: src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs (hunk 7 above)
Comment (nschwerzler) [reply src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs]: No errors when file already exists and refreshes with new file.

File: src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs (hunk 8 above)
Comment (nschwerzler) [reply src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs]: Resolved

File: src/NuGet.Core/NuGet.Commands/CommandArgs/PackArgs.cs (hunk 10 above)
Comment (nschwerzler) [reply src/NuGet.Core/NuGet.Commands/CommandArgs/PackArgs.cs]: Resolved. Updated name. Normal nupkg and symbols are no longer generated.

File: src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs (hunk 12 above)
Comment (nschwerzler) [reply src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs]: Resolved

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPackCommandTest.cs (hunk 9 above)
Comment (nschwerzler) [reply test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPackCommandTest.cs]: Resolved

File: src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs (hunk 11 above)
Comment (rohit21agrawal) [reply src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs]: by installing the package I mean when InstallPackageToV3Feed is set
